add useSSR option to enable componentId based on file path

fabyo82 opened the pull request:
add styledIdentifiers to enable replacing wrapped styled components (as styled-components-ts)

implements #20

Igorbek commented on src/createTransformer.ts:
```diff
@@ -94,6 +98,31 @@ export function createTransformer({ getDisplayName = defaultGetDisplayName }: Pa
     }
 
+    const separatorRegExp = new RegExp(`\\${path.sep}`, 'g');
+    const findModuleRoot = (filename: string): string | null => {
```
the logic that finds module root seems to rely on very specific layout. Sometimes package.json may be in subfolders or project may not have it at all. To make it more generic we may accept root folder as an option or have consumer-specified function with custom resolution.

fabyo82 replied:
Fine for me to have a root folder option. The babel plugin adopted this solution, but it's true that the find-root module accepts an additional function in order to be more generic

Igorbek commented on src/createTransformer.ts:
```diff
@@ -94,6 +98,31 @@ export function createTransformer({ getDisplayName = defaultGetDisplayName }: Pa
+            const fileName = node.getSourceFile().fileName;
+            const moduleRoot = findModuleRoot(fileName)
+            const filePath = moduleRoot ? path.relative(moduleRoot, fileName).replace(separatorRegExp, '/') : '';
+            return 'sc-' + hash(filePath);
```
The hash is calculated only based on the file name. Does Babel plugin also do this the same way? What the module has multiple components? Would they all have the same id?

Igorbek replied:
It seems that Babel plugin does it differently:

https://github.com/styled-components/babel-plugin-styled-components/blob/a092bc8e40d4d4b0c85f80b33e36eb093951a117/src/visitors/displayNameAndId.js#L131-L134

fabyo82 replied:
Currently it's working because the displayName is set as well, so different components in the same file would have a different displayName but same componentId.
We could make the displayName optional for the production use, then add a counter similar to the babel plugin.
If I keep a counter as a variable inside the transformer I faced some issues in the watching mode. It may be that the server doesn't refresh the same number of times of the client.
Babel uses a state, but I'm not sure if it's possible using the transformer and if it would really fix the issue I faced, any idea?
As alternative we could hash as componentId the file name together with the node name (the same used by default to set the displayName)

Igorbek commented on src/createTransformer.ts:
```diff
@@ -102,15 +131,25 @@ export function createTransformer({ getDisplayName = defaultGetDisplayName }: Pa
                 const displayName = getDisplayNameFromNode(node.parent.parent);
+                let componentId;
+                if(useSSR){
+                    componentId = getIdFromNode(node.parent.parent);
```
does it always have to be that SSR ties it with component id generation? I'm curious if component id is useful for anything else.

fabyo82 replied:
good question, anyway it would make sense to allow customize the componentId regardless the useSSR flag

fabyo82 commented:
Thank you for the feedback @Igorbek, I'd be fine if you apply the changes, or I could do it as well if you prefer (depending on my availability ;-))
I'd have some concerns about the options as functions, as I'm using the ttypescript module to compile the server, and it allows to set the transformer options directly inside the tsconfig.json, so a function would require an additional wrapper.
What if we try to keep the options as close as possible to the bable plugin? We can review together the new contract before start implementing the changes.
About the unit tests I wasn't sure what's the best way to add the new test cases, so I let you add them or please propose a solution. 
Thanks

fabyo82 commented:
hello @Igorbek , sorry to bother you, I'd like to advance on this topic. Please let me know if you are still interested in this PR, or I'll handle it differently. Of course your feedback would be much appreciated

Igorbek commented:
Oh, sorry @fabyo82 I actually thought you were going to make changes and I forgot to follow up on this.
If you're ok to update, I would like you first propose API change so we can agreed upon and move forward with implementing it.

Igorbek commented on package.json:
```diff
@@ -2,7 +2,7 @@
     "version": "0.0.6",
     "description": "TypeScript transformer for improving the debugging experience of styled-components",
-    "main": "dist/index.js",
+    "main": "distor/index.js",
```
leftover?

Igorbek commented on src/createTransformer.ts:
```diff
@@ -18,9 +21,9 @@ import {Options} from './models/Options';
  * styledFunction.attrs(attributes)
 */
-function isStyledFunction(node: ts.Node): boolean {
+function isStyledFunction(node: ts.Node, styledIdentifiers: string[]): boolean {
```
As we bring more options, we'll bloat this function more. I suggest to add context and pass it around.

```ts
interface TransformContext {
  readonly options: Options;
}
```

and just use it directly like this:

```ts
function isStyledFunction(node: ts.Node, { options: { styledIdentifiers } }: TransformContext): boolean
```

Igorbek replied:
Let me take it back. That should be fine since it is isolated responsibility, and it is ok to accept specific parameters.

Igorbek commented on src/createTransformer.ts:
```diff
@@ -94,6 +98,31 @@ export function createTransformer({ getDisplayName = defaultGetDisplayName }: Pa
+            const fileName = node.getSourceFile().fileName;
+            const moduleRoot = findModuleRoot(fileName)
+            const filePath = moduleRoot ? path.relative(moduleRoot, fileName).replace(separatorRegExp, '/') : '';
+            return 'ssr-' + hash(`${getDisplayNameFromNode(node)}${filePath}`);
```
It seems that babel plugin uses hash based on relative path **and** component position. That makes sense as we need to distinguish components within a file.

fabyo82 replied:
do you have any idea how to calculate the component position? babel keeps a counter inside the visitor state, I couldn't find a similar option for the typescript transformer.

Instead of the position I added the getDisplayNameFromNode (thus bindingName by default). This would generate different componentIds for components in the same file

Igorbek commented on src/models/Options.ts:
```diff
@@ -5,4 +5,10 @@ export interface Options {
+
+    ssr: boolean;
+    displayName: boolean;
+    rootCheck: string;
```
could you please give a brief description (what they mean and do) of each new option here

fabyo82 replied:
I propose to keep the same names used by the babel plugin. For this reason there's a boolean "displayName". For the same reason I didn't group them inside the "experimental" section

karolfalkiewicz commented:
Would be nice to add with this PR the possibility of same displayName behavior as babel plugin (no more styled.button, better to have actual component name here).

Igorbek commented:
@karolfalkiewicz could you explain what is exactly the difference in `displayName` you referring to? Actual component name has been supported by this plugin from the beginning.

fabyo82 commented:
Hi @Igorbek ,

I updated my PR following some of your suggestions. 
My main goal is to provide a working example for a Typescript SSR styled-components application.
As you pointed out, I have a doubt about the case having multiple components in the same file: my example works, but  I adopted a different solution compared to the babel plugin, thus may not work under certain conditions.
Also I don't know how would you prefer you to extend the unit tests, given that I added a few parameters.
Let me know if you prefer to update yourself the PR
thanks

Igorbek commented:
@fabyo82 please see my additions in my pr yo your pr: https://github.com/fabyo82/typescript-plugin-styled-components/pull/2

karolfalkiewicz commented:
Right now in React devtools I see components like `styled.h1` etc. I would like to see actual variable name here, like `Title`. Additional thing is to add the actual variable name to styled-components generated classes, the same way as library using babel do this. If this possible right now then ok, but couldn't find any way to achieve this (tried using options but with no luck).

Igorbek commented:
@karolfalkiewicz what you're referring to is a core functionality of the plugin. If you experiencing difficulties making it work, it might be worth filing a separate issue as it sounds like a bug. This PR seems unrelated to the issue you have and I would ask to not bloat off-topic discussion here. If you file an issue, please consider adding a reproduction repository to simplify investigation.

Igorbek commented:
Thank you @fabyo82 for the contribution. I'll need 1-2 days to cleanup a little bit and publish a new version.

fabyo82 commented:
thank you @Igorbek , we should update the readme as well

Igorbek commented:
sure, added to checklist #25
